src/aruku/walking/process/kinematic.cpp

```diff
@@ -0,0 +1,729 @@
+    cos(angles[JointId::RIGHT_HIP_YAW + index_addition]) *
+    sin(angles[JointId::RIGHT_HIP_ROLL + index_addition]) * vector.y;
+  double s = (k * n + l * m) / (k * k + l * l);
+  double t = (n - k * s) / l;
+
+  double theta_result = atan_result;
+  atan_result = atan2(s, t);
+  if (std::isinf(atan_result)) {
+    return false;
+  }
+  angles[JointId::RIGHT_HIP_PITCH + index_addition] = atan_result;
+  angles[JointId::RIGHT_ANKLE_PITCH + index_addition] = theta_result -
+    angles[JointId::RIGHT_KNEE + index_addition] -
+    angles[JointId::RIGHT_HIP_PITCH + index_addition];
+
+  return true;
+}
+
+void Kinematic::update_times()
+{
+  double dsp_comp = fabs(m_x_move_amplitude) * dsp_comp_ratio * 0.001;
+
+  m_period_time = period_time - (fabs(m_x_move_amplitude) * period_comp_ratio);
+
+  m_dsp_ratio = dsp_ratio + dsp_comp;
+  m_ssp_ratio = 1 - m_dsp_ratio;
+
+  m_x_swap_period_time = m_period_time / 2;
+  m_x_move_period_time = m_period_time * m_ssp_ratio;
+  m_y_swap_period_time = m_period_time;
+  m_y_move_period_time = m_period_time * m_ssp_ratio;
+  m_z_swap_period_time = m_period_time / 2;
+  m_z_move_period_time = m_period_time * m_ssp_ratio / 2;
+  m_a_move_period_time = m_period_time * m_ssp_ratio;
+
+  m_ssp_time = m_period_time * m_ssp_ratio;
+  m_ssp_time_start_l = (1 - m_ssp_ratio) * m_period_time / 4;
+  m_ssp_time_end_l = (1 + m_ssp_ratio) * m_period_time / 4;
+  m_ssp_time_start_r = (3 - m_ssp_ratio) * m_period_time / 4;
+  m_ssp_time_End_r = (3 + m_ssp_ratio) * m_period_time / 4;
+
+  m_phase_time1 = (m_ssp_time_start_l + m_ssp_time_end_l) / 2;
+  m_phase_time2 = (m_ssp_time_end_l + m_ssp_time_start_r) / 2;
+  m_phase_time3 = (m_ssp_time_start_r + m_ssp_time_End_r) / 2;
+
+  m_arm_swing_gain = arm_swing_gain;
+
+  if (m_x_move_amplitude > 0) {
+    hip_comp = m_x_move_amplitude * forward_hip_comp_ratio;
+  } else {
+    hip_comp = m_x_move_amplitude * backward_hip_comp_ratio;
+  }
+
+  foot_comp = fabs(m_x_move_amplitude) * foot_comp_ratio;
+}
+
+void Kinematic::update_move_amplitude()
+{
+  double x_input = x_move;
+  double y_input = y_move * 0.5;
+  double a_input = a_move;
+
+  if (m_z_move_amplitude < (z_move * 0.45)) {
+    x_input = 0.0;
+    y_input = 0.0;
+    a_input = 0.0;
+  }
+
+  m_x_move_amplitude = keisan::smooth(m_x_move_amplitude, x_input, move_accel_ratio);
+  m_y_move_amplitude = keisan::smooth(m_y_move_amplitude, y_input, move_accel_ratio);
+  m_y_move_amplitude_shift = fabs(m_y_move_amplitude);
+  m_y_swap_amplitude = y_swap_amplitude + m_y_move_amplitude_shift * 0.04;
+
+  if (m_ctrl_running) {
+    m_z_move_amplitude = keisan::smooth(
+      m_z_move_amplitude, (z_move + foot_comp) * 0.5, foot_accel_ratio);
+  } else {
+    m_z_move_amplitude = 0.0;
+  }
+
+  m_z_move_amplitude_shift = m_z_move_amplitude;
+  m_z_swap_amplitude = z_swap_amplitude;
+
+  if (!a_move_aim_on) {
+    m_a_move_amplitude = keisan::smooth(
+      m_a_move_amplitude, keisan::make_degree(a_input).radian() / 2, move_accel_ratio);
```

Comment:
i suggest to use `keisan::Angle` for `a_input` in the first place instead.

Reply:
done